service/history/replicationDLQHandler.go

```diff
@@ -0,0 +1,227 @@
+			sourceCluster string,
+			lastMessageID int64,
+			pageSize int,
+			pageToken []byte,
+		) ([]byte, error)
+	}
+
+	replicationDLQHandlerImpl struct {
+		replicationTaskExecutor replicationTaskExecutor
+		shard                   ShardContext
+		logger                  log.Logger
+	}
+)
+
+func newReplicationDLQHandler(
+	shard ShardContext,
+	replicationTaskExecutor replicationTaskExecutor,
+) replicationDLQHandler {
+
+	return &replicationDLQHandlerImpl{
+		shard:                   shard,
+		replicationTaskExecutor: replicationTaskExecutor,
+		logger:                  shard.GetLogger(),
+	}
+}
+
+func (r *replicationDLQHandlerImpl) readMessages(
+	ctx context.Context,
+	sourceCluster string,
+	lastMessageID int64,
+	pageSize int,
+	pageToken []byte,
+) ([]*replicator.ReplicationTask, []byte, error) {
+
+	ackLevel := r.shard.GetReplicatorDLQAckLevel(sourceCluster)
+	resp, err := r.shard.GetExecutionManager().GetReplicationTasksFromDLQ(&persistence.GetReplicationTasksFromDLQRequest{
+		SourceClusterName: sourceCluster,
+		GetReplicationTasksRequest: persistence.GetReplicationTasksRequest{
+			ReadLevel:     ackLevel,
+			MaxReadLevel:  lastMessageID,
+			BatchSize:     pageSize,
+			NextPageToken: pageToken,
+		},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	remoteAdminClient := r.shard.GetService().GetClientBean().GetRemoteAdminClient(sourceCluster)
+	taskInfo := make([]*replicator.ReplicationTaskInfo, len(resp.Tasks))
+	for _, task := range resp.Tasks {
+		taskInfo = append(taskInfo, &replicator.ReplicationTaskInfo{
+			DomainID:     common.StringPtr(task.GetDomainID()),
+			WorkflowID:   common.StringPtr(task.GetWorkflowID()),
+			RunID:        common.StringPtr(task.GetRunID()),
+			TaskType:     common.Int16Ptr(int16(task.GetTaskType())),
+			TaskID:       common.Int64Ptr(task.GetTaskID()),
+			Version:      common.Int64Ptr(task.GetVersion()),
+			FirstEventID: common.Int64Ptr(task.FirstEventID),
+			NextEventID:  common.Int64Ptr(task.NextEventID),
+			ScheduledID:  common.Int64Ptr(task.ScheduledID),
+		})
+	}
+	dlqResponse, err := remoteAdminClient.GetDLQReplicationMessages(
+		ctx,
+		&replicator.GetDLQReplicationMessagesRequest{
+			TaskInfos: taskInfo,
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return dlqResponse.ReplicationTasks, resp.NextPageToken, nil
+}
+
+func (r *replicationDLQHandlerImpl) purgeMessages(
+	sourceCluster string,
+	lastMessageID int64,
+) error {
+
+	ackLevel := r.shard.GetReplicatorDLQAckLevel(sourceCluster)
+	err := r.shard.GetExecutionManager().RangeDeleteReplicationTaskFromDLQ(
+		&persistence.RangeDeleteReplicationTaskFromDLQRequest{
+			SourceClusterName:    sourceCluster,
+			ExclusiveBeginTaskID: ackLevel,
+			InclusiveEndTaskID:   lastMessageID,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if err = r.shard.UpdateReplicatorDLQAckLevel(
+		sourceCluster,
+		lastMessageID,
+	); err != nil {
+		r.logger.Error("Failed to purge history replication message", tag.Error(err))
```

Comment:
missing return statement. Is this intentional ?

Reply:
Yes